<a href="https://colab.research.google.com/github/vlx300/kb_colab/blob/master/Python_Pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#**Python Pandas Data Exploration**
![alt text](https://ageng.files.wordpress.com/2008/06/kung_fu_panda_xl_021.jpg)

#**Mount Google Drive**

In [0]:
 from google.colab import drive  # mount google drive #
drive.mount('/content/drive')

###**Append sys.path for Google Drive**

In [0]:
import sys
sys.path.append("/content/drive/My Drive")

##**Grabbing the dataset**: 
**Five-thirty-Eight: NBA Results**  (https://data.fivethirtyeight.com/)

In [0]:
import requests   # import requestas library

download_url = "https://raw.githubusercontent.com/fivethirtyeight/data/master/nba-elo/nbaallelo.csv"  #data source stored in variable download_url 
target_csv_path = "/content/drive/My Drive/nba_all_elo.csv"  # valible to store and use target file path #

In [0]:
response = requests.get(download_url)
response.raise_for_status()  # check that the request was successful 
with open(target_csv_path, "wb") as f:
  f.write(response.content)
print("download ready")

download ready


#**Python Pandas: Taking a closer look at the Data**

as you can see above, we import pandas as pd (alias), then we utilize read_csv() to read in the dataset, and store as a Pandas dataframe in a varible called nba. 

**Note:** Is your data NOT cvs format? the Pandas libray contains several other similar functions (read_json(), read_html(), read_sql_table()

In [0]:
import pandas as pd
nba = pd.read_csv("/content/drive/My Drive/nba_all_elo.csv") # store as pandas dataframe object in a variable called nba
type(nba)

pandas.core.frame.DataFrame

You can utilize the built-in len() function to determine the numbr of rows

In [0]:
len(nba)  # how much data does nba contain (# of rows in dataset)

126314

In [0]:
nba.shape # you can see it dimensionality .TUPLE containing the number of rows and columns

(126314, 23)

Now we know the dimensions of our dataset, How can we be sure that the dataset actually contains basketball stats? Take a look a tthe 1st five rows with head()

In [0]:
nba.head()  # displays first 5 rows in dataset to view the contents

You can configure pandas to display all columns like this

In [0]:
pd.set_option("display.max.columns", None)  # allows you to see ALL columns in dataset 

Also change the number of decimal places to display.

In [0]:
pd.set_option("display.precision", 1)  # only show 1 decimal place # 

In [0]:
nba.head()  # display first five rows

gameorder       game_id lg_id  _iscopy  year_id  date_game  seasongame  \
0          1  194611010TRH   NBA        0     1947  11/1/1946           1   
1          1  194611010TRH   NBA        1     1947  11/1/1946           1   
2          2  194611020CHS   NBA        0     1947  11/2/1946           1   
3          2  194611020CHS   NBA        1     1947  11/2/1946           2   
4          3  194611020DTF   NBA        0     1947  11/2/1946           1   

   is_playoffs team_id  fran_id  pts   elo_i   elo_n  win_equiv opp_id  \
0            0     TRH  Huskies   66  1300.0  1293.3       40.3    NYK   
1            0     NYK   Knicks   68  1300.0  1306.7       41.7    TRH   
2            0     CHS    Stags   63  1300.0  1309.7       42.0    NYK   
3            0     NYK   Knicks   47  1306.7  1297.1       40.7    CHS   
4            0     DTF  Falcons   33  1300.0  1279.6       38.9    WSC   

   opp_fran  opp_pts  opp_elo_i  opp_elo_n game_location game_result  \
0    Knicks       68     1300.0     1306.7             H           L   
1   Huskies       66     1300.0     1293.3             A           W   
2    Knicks       47     1306.7     1297.1             H           W   
3     Stags       63     1300.0     1309.7             A           L   
4  Capitols       50     1300.0     1320.4             H           L   

   forecast notes  
0       0.6   NaN  
1       0.4   NaN  
2       0.6   NaN  
3       0.4   NaN  
4       0.6   NaN

In [0]:
nba.tail()   # display last five rows

gameorder       game_id lg_id  _iscopy  year_id  date_game  \
126309      63155  201506110CLE   NBA        0     2015  6/11/2015   
126310      63156  201506140GSW   NBA        0     2015  6/14/2015   
126311      63156  201506140GSW   NBA        1     2015  6/14/2015   
126312      63157  201506170CLE   NBA        0     2015  6/16/2015   
126313      63157  201506170CLE   NBA        1     2015  6/16/2015   

        seasongame  is_playoffs team_id    fran_id  pts   elo_i   elo_n  \
126309         100            1     CLE  Cavaliers   82  1723.4  1704.4   
126310         102            1     GSW   Warriors  104  1810.0  1813.6   
126311         101            1     CLE  Cavaliers   91  1704.4  1700.7   
126312         102            1     CLE  Cavaliers   97  1700.7  1692.1   
126313         103            1     GSW   Warriors  105  1813.6  1822.3   

        win_equiv opp_id   opp_fran  opp_pts  opp_elo_i  opp_elo_n  \
126309       60.3    GSW   Warriors      103     1791.0     1810.0   
126310       68.0    CLE  Cavaliers       91     1704.4     1700.7   
126311       60.0    GSW   Warriors      104     1810.0     1813.6   
126312       59.3    GSW   Warriors      105     1813.6     1822.3   
126313       68.5    CLE  Cavaliers       97     1700.7     1692.1   

       game_location game_result  forecast notes  
126309             H           L       0.5   NaN  
126310             H           W       0.8   NaN  
126311             A           L       0.2   NaN  
126312             H           L       0.5   NaN  
126313             A           W       0.5   NaN

In [0]:
nba.tail(7) # displays the last 3 rows of the dataset #

gameorder       game_id lg_id  _iscopy  year_id  date_game  \
126307      63154  201506090CLE   NBA        0     2015   6/9/2015   
126308      63155  201506110CLE   NBA        1     2015  6/11/2015   
126309      63155  201506110CLE   NBA        0     2015  6/11/2015   
126310      63156  201506140GSW   NBA        0     2015  6/14/2015   
126311      63156  201506140GSW   NBA        1     2015  6/14/2015   
126312      63157  201506170CLE   NBA        0     2015  6/16/2015   
126313      63157  201506170CLE   NBA        1     2015  6/16/2015   

        seasongame  is_playoffs team_id    fran_id  pts   elo_i   elo_n  \
126307          99            1     CLE  Cavaliers   96  1716.9  1723.4   
126308         101            1     GSW   Warriors  103  1791.0  1810.0   
126309         100            1     CLE  Cavaliers   82  1723.4  1704.4   
126310         102            1     GSW   Warriors  104  1810.0  1813.6   
126311         101            1     CLE  Cavaliers   91  1704.4  1700.7   
126312         102            1     CLE  Cavaliers   97  1700.7  1692.1   
126313         103            1     GSW   Warriors  105  1813.6  1822.3   

        win_equiv opp_id   opp_fran  opp_pts  opp_elo_i  opp_elo_n  \
126307       61.8    GSW   Warriors       91     1797.5     1791.0   
126308       67.8    CLE  Cavaliers       82     1723.4     1704.4   
126309       60.3    GSW   Warriors      103     1791.0     1810.0   
126310       68.0    CLE  Cavaliers       91     1704.4     1700.7   
126311       60.0    GSW   Warriors      104     1810.0     1813.6   
126312       59.3    GSW   Warriors      105     1813.6     1822.3   
126313       68.5    CLE  Cavaliers       97     1700.7     1692.1   

       game_location game_result  forecast notes  
126307             H           W       0.5   NaN  
126308             A           W       0.5   NaN  
126309             H           L       0.5   NaN  
126310             H           W       0.8   NaN  
126311             A           L       0.2   NaN  
126312             H           L       0.5   NaN  
126313             A           W       0.5   NaN

##**Getting to know the data**
![alt text](https://miro.medium.com/max/403/1*TUw8sTRFiMgV0KRyUzjWmQ.jpeg)


you imported a CSV file, with the pandas libray and had a 1st look a the contents of your dataset, so far you only seen the size of the dataset and the first/last few rows. 

**Now lets examine the data more systematically**

**Displaying the Data Types**

The first steps in getting to know your data is having a understanding of the **data-types** it contains.  While you can place anything into a list, the columns of a **Dataframe** contain values of a specific data type. When you compare Pandas and Python data structures you see that this behavior makes Pandas much faster.  

you can display ALL columns and their data types with .info()

In [0]:
nba.info()  # this will display ALL of the columns in your dataset, and the TYPE of data each column contains 
# Here you can see the following data types : int64, float64 and object.  Pandas uses numpy to work with these data types 
# the object data type is a special one. "Catch-all" for the data types that python doesnt recognize.   "strings" in practice!

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 126314 entries, 0 to 126313
Data columns (total 23 columns):
gameorder        126314 non-null int64
game_id          126314 non-null object
lg_id            126314 non-null object
_iscopy          126314 non-null int64
year_id          126314 non-null int64
date_game        126314 non-null object
seasongame       126314 non-null int64
is_playoffs      126314 non-null int64
team_id          126314 non-null object
fran_id          126314 non-null object
pts              126314 non-null int64
elo_i            126314 non-null float64
elo_n            126314 non-null float64
win_equiv        126314 non-null float64
opp_id           126314 non-null object
opp_fran         126314 non-null object
opp_pts          126314 non-null int64
opp_elo_i        126314 non-null float64
opp_elo_n        126314 non-null float64
game_location    126314 non-null object
game_result      126314 non-null object
forecast         126314 non-null float64
notes     

You see a list of ALL the columns in your dataset and the type of data each column contains.  You can see three (3)**bold text** data types: **int64, float64**, and **object**.  Pandas use NumPy library to work with these types Later you will the more complex "categorical" data type which the Pandas library implements itself. 


the **object** data type is a special one. The Object data type is a "*catch-all*" for columns that Pandas doesnt recognize as any pother specific data type.  In practice,  it means that the values in the columns are strings.

although you can store arbitrary python objects in the object data type, there are some drawbacks

*   Strange values in a object column can harm pandas performance
*   interoperaility with other libraries can also be affected. 



##**Showing Basic Statistics**

Now that you've seen what datatypes are in your dataset, lets get an overview of the values in each column. You can do this with .describe()

In [0]:
nba.describe() #show basics stata for all numerical (by default) columns #

gameorder   _iscopy   year_id  seasongame  is_playoffs       pts  \
count   126314.0  126314.0  126314.0    126314.0      1.3e+05  126314.0   
mean     31579.0       0.5    1988.2        43.5      6.4e-02     102.7   
std      18231.9       0.5      17.6        25.4      2.4e-01      14.8   
min          1.0       0.0    1947.0         1.0      0.0e+00       0.0   
25%      15790.0       0.0    1975.0        22.0      0.0e+00      93.0   
50%      31579.0       0.5    1990.0        43.0      0.0e+00     103.0   
75%      47368.0       1.0    2003.0        65.0      0.0e+00     112.0   
max      63157.0       1.0    2015.0       108.0      1.0e+00     186.0   

          elo_i     elo_n  win_equiv   opp_pts  opp_elo_i  opp_elo_n  forecast  
count  126314.0  126314.0   126314.0  126314.0   126314.0   126314.0   1.3e+05  
mean     1495.2    1495.2       41.7     102.7     1495.2     1495.2   5.0e-01  
std       112.1     112.5       10.6      14.8      112.1      112.5   2.2e-01  
min      1091.6    1085.8       10.2       0.0     1091.6     1085.8   2.0e-02  
25%      1417.2    1417.0       34.1      93.0     1417.2     1417.0   3.3e-01  
50%      1500.9    1501.0       42.1     103.0     1500.9     1501.0   5.0e-01  
75%      1576.1    1576.3       49.6     112.0     1576.1     1576.3   6.7e-01  
max      1853.1    1853.1       71.1     186.0     1853.1     1853.1   9.8e-01

This function show you some more basic descriptive statisitics for all numerical columns. 

NOTE: .describe() only analyzes numerical columns by default. But you can provide other data types  if you use the "include" parameter

In [0]:
import numpy as np
nba.describe(include=np.object)  # use include parameter for non int64/float64 values #

game_id   lg_id  date_game team_id fran_id  opp_id opp_fran  \
count         126314  126314     126314  126314  126314  126314   126314   
unique         63157       2      12426     104      53     104       53   
top     199711150CHH     NBA  4/17/2013     BOS  Lakers     BOS   Lakers   
freq               2  118016         30    5997    6024    5997     6024   

       game_location game_result           notes  
count         126314      126314            5424  
unique             3           2             231  
top                A           L  at New York NY  
freq           63138       63157             440

.describe wont try to calculate a *MEAN or Standard Deviation* for the object columns, since they mostly include text strings but still will provide some descriptivie statistics. 

Take a look at the team_id and Fran_id columns. the data set contains 104 team IDs but only 53 franchise IDs. Furthermore, the most frequent team ID is BOS (Boston), and the most frequent franchise is lakers.   How is this possible?

Lets explore a bit more...

##**Exploratory Data Analysis**

exploratatory data analysis can help you answer questions about your dataset. for example, you can examine how often values occur in a specific column

In [0]:
nba["team_id"].value_counts()

BOS    5997
NYK    5769
LAL    5078
DET    4985
PHI    4533
       ... 
TRH      60
INJ      60
PIT      60
DTF      60
SDS      11
Name: team_id, Length: 104, dtype: int64

In [0]:
nba["fran_id"].value_counts()

Lakers          6024
Celtics         5997
Knicks          5769
Warriors        5657
Pistons         5650
Sixers          5644
Hawks           5572
Kings           5475
Wizards         4582
Spurs           4309
Bulls           4307
Pacers          4227
Thunder         4178
Rockets         4154
Nuggets         4120
Nets            4106
Suns            4080
Bucks           4034
Trailblazers    3870
Cavaliers       3810
Clippers        3733
Jazz            3555
Mavericks       3013
Heat            2371
Pelicans        2254
Magic           2207
Timberwolves    2131
Grizzlies       1657
Raptors         1634
Hornets          894
Colonels         846
Squires          799
Spirits          777
Stars            756
Sounds           697
Baltimore        467
Floridians       440
Condors          430
Capitols         291
Olympians        282
Sails            274
Stags            260
Bombers          249
Steamrollers     168
Packers           72
Redskins          65
Rebels            63
Waterloo     

It seems that a team named "lakers" played 6024 games but only 5078 of those were played by the Los Angeles lakers.  who is the other lakers team?

In [0]:
nba.loc[nba["fran_id"] == "Lakers", "team_id"].value_counts()

LAL    5078
MNL     946
Name: team_id, dtype: int64

Las Angeles lakers (LAL) and Minneapolis Lakers (MNL) played 946 games. We can even find out when those games were played. 

In [0]:
nba.loc[nba["team_id"] == "MNL", "date_game"].min()

'1/1/1949'

In [0]:
nba.loc[nba["team_id"] == "MNL", "date_game"].max()

'4/9/1959'

In [0]:
nba.loc[nba["team_id"] == "MNL", "date_game"].agg(("min", "max"))

min    1/1/1949
max    4/9/1959
Name: date_game, dtype: object

Looks like the Mineanapolis lakers played between the years 1949 and 1959. This explains why we didnt recognize the team. Youve also found out why the Boston Celtics (BOS)played the most games in the dataset.  Lets analyze their history also a little bit.  Let find out how many points the boston celtics scored during all matches contained in the dataset. 

In [0]:
nba.loc[nba["team_id"] == "BOS", "pts"].sum()

626484

The Boston Celtics scored a total of 626,484 points

#**Pandas Data strucutures**

###**Understanding Series Objects**

Pyhons most basic data structure is a list. Which is also a great starting point for getting to know pandas.series Objects. 

In [0]:
revenues = pd.Series([5555, 7000, 1980])

In [0]:
revenues

0    5555
1    7000
2    1980
dtype: int64

you've used the list [5555, 7000, 1980] to create a series object called revenues. a series object wraps two components:

1.   A sequence of **values** (5555,7000,1980)
2.   A sequence of identifiers, which is the index (0,1,2)

you can acces these components with .values and .index, respectively



In [0]:
revenues.values  # returns the values in the series

array([5555, 7000, 1980])

In [0]:
revenues.index  # returns the positional index

RangeIndex(start=0, stop=3, step=1)



NOTE: It may be of interest to note that the values of a series object are actually N-Dimensional arrays.  See NumPy for more details (https://numpy.org/)

In [0]:
type(revenues.values)

numpy.ndarray

while pandas builds on Numpy you can explore the ins and outs of your dataset with Pandas alone. **A significant difference is in their indexin**g.

NOTE: Just like a Numpy array, a pandas series also has a integer index that "implicitly" defined, this implicit index **indicates the elements positoin in a series**. A series can also have a "Arbitrary type of Index" you can **think of this explicit index as labels for specific rows**.  

In [0]:
city_revenues = pd.Series([4200, 8000, 6500], index=["Amsterdam", "Toronto", "Tokyo"])

In [0]:
city_revenues

Amsterdam    4200
Toronto      8000
Tokyo        6500
dtype: int64

Here the index is a list of city names represented as strings. you may have notices that python dictionaries uses string indices as well {key:value}. and this is a handly analogy ot keep in mind. you can use the code blocks above to distringuish between the two types of series:

1. revenues: This series behanves much like a Python listbecaause it only has a positional index. 

2. city_revenues: this series acta like a python dictionary because it features both a positional and label index. 

Here is how to construct a series with a label index from a python dictionary:

In [0]:
city_employee_count = pd.Series({"Amsterdam":5, "Tokyo": 8})

In [0]:
city_employee_count

Amsterdam    5
Tokyo        8
dtype: int64

the dictionary keys become the index and the values become become the series values. just like dictionaries, series also support .keys() inthe keyword. 

In [0]:
city_employee_count.keys()  # values() wont work with nmupy array

Index(['Amsterdam', 'Tokyo'], dtype='object')

In [0]:
"Tokyo" in city_employee_count

True

In [0]:
"New Jersey" in city_employee_count

False

you can use these methods to answer questions about your data set quickly. 

##**Understanding DataFrame objects**

While a series is a pretty powerful data structure, it has it limitations, Exp, you can *only store one attribute per key*. As you have seen with the NBA dataset.  there are 23 columns. Pandas has  more to offer via the DataFrame. this data structure is sequence of Series object that share the same index. 

you already have two series objects with cities as keys

In [0]:
city_revenues

Amsterdam    4200
Toronto      8000
Tokyo        6500
dtype: int64

In [0]:
city_employee_count

Amsterdam    5
Tokyo        8
dtype: int64

you can combine these objext into a data frame by providing a dictionary in the constructor. the dictionary keys will become the column names, and the values conmtain thr Series objects: 

In [0]:
city_data = pd.DataFrame({"revenue": city_revenues, "employee_count": city_revenues})

In [0]:
city_data

revenue  employee_count
Amsterdam     4200            4200
Toronto       8000            8000
Tokyo         6500            6500

In [0]:
city_data.index

Index(['Amsterdam', 'Toronto', 'Tokyo'], dtype='object')

just like in a Series, a dataframe also stores values in a NumPy array

In [0]:
city_data.values

array([[4200, 4200],
       [8000, 8000],
       [6500, 6500]])

you can also refer to the 2 dimensions of the Dataframe as **axes**. 

In [0]:
city_data.axes

[Index(['Amsterdam', 'Toronto', 'Tokyo'], dtype='object'),
 Index(['revenue', 'employee_count'], dtype='object')]

In [0]:
city_data.axes[0]

Index(['Amsterdam', 'Toronto', 'Tokyo'], dtype='object')

In [0]:
city_data.axes[1]

Index(['revenue', 'employee_count'], dtype='object')

the axis marked with [0] is the row index, the axes ,makred with [1] is the column index. A dtaframe is also a dictionary like structure, so it also supports .keys() and the IN keyword. However for a dataframe, these dont relate to the index, but to the columns


In [0]:
city_data.keys()

Index(['revenue', 'employee_count'], dtype='object')

In [0]:
"Amsterdam" in city_data

False

In [0]:
"revenue" in city_data

True

As you use these methods to answer questions about your dataset, keep in mind whether your working with a Series or a DataFrame so your interpretation is accurate 

#**Accessing Series Elements** 

in the section above, your created a Pandas series based upon a python list, and compared the two data strcutures, Youve see how  a seroies objectis similar to lists anmdditionairies ain several ways. a further similarity is that can use the indexing operator ([])  for Series as well. 

you have also learned how to use two Pandas Specifc acess methods: 


1.   .loc
2.   .iloc

you'll also learn that these data access methods can be much more readable than the indexiing operator. 

##**Indexing Operator**

recal a series has two indexes: 
1. A Positional or implicit index - whixh is always the range index, 
2. a Label oe explicit index, which can contain any hashable objects.  


In [0]:
city_revenues

Amsterdam    4200
Toronto      8000
Tokyo        6500
dtype: int64

You can conveniently access the values in a series with both **label** and **positional indexes**

In [0]:
city_revenues["Toronto"]

8000

In [0]:
city_revenues[2]

6500

You can also utilize negative indices and slices as well

In [0]:
city_revenues[-1]

6500

In [0]:
city_revenues[:-2]

Amsterdam    4200
dtype: int64

##**Pandas Python library Data access methods** 

**.loc = label index**

**.iloc = positional index**

these data access methods are much more readable

![alt text](https://files.realpython.com/media/iloc_vs_loc_80_border20.d5280f475f4e.png)

In [0]:
colors = pd.Series(["red", "purple", "blue", "green", "yellow"], index=[1, 2, 3, 5, 8 ])

In [0]:
colors

1       red
2    purple
3      blue
5     green
8    yellow
dtype: object

In [0]:
colors.loc[1]

'red'

In [0]:
colors.iloc[1]

'purple'

In [0]:
colors.iloc[:3]  # can do slices 

1       red
2    purple
3      blue
dtype: object

##Slicing Note:
 .loc and iloc support th features you would expect from indexiing operations, However, the data acess methods have an important difference,  .iloc excludes the cosing element while .loc includes it. 

In [0]:
colors.loc[:4]

1       red
2    purple
3      blue
dtype: object

In [0]:
colors.iloc[:4]

1       red
2    purple
3      blue
5     green
dtype: object

#**Accessing dataframe elements**

since a dataframe consists of series objects, you can use the very same tools to access it elements. the crucial difference is the additional dimensions of a dataframe. You will use the index operator for the columns and the access methods (.loc and iloc) for the rows



##Using the index operator

If you think of a dataframe as a dictionary whose values are a series, the it makes sense that you can access its columns with the indexing operator

In [0]:
city_data["revenue"]

Amsterdam    4200
Toronto      8000
Tokyo        6500
Name: revenue, dtype: int64

Here, you use the indexing operator to select the column labeled revenue.  
NOTE: if the column name is a string, then you can yse the attribute-style accessing with .dot notation as well

In [0]:
city_data.revenue

Amsterdam    4200
Toronto      8000
Tokyo        6500
Name: revenue, dtype: int64

there is one situation where accessing Dataframe elements may not work and may lead to surprises. this is when a column name coincides with a dataframe attribute or method name

In [0]:
toys = pd.DataFrame([
                     {"name": "ball", "shape": "sphere"},
                     {"name": "Rubik's cube", "'shape": "cube"}
                     ])

In [0]:
toys["shape"]

0    sphere
1       NaN
Name: shape, dtype: object

In [0]:
toys.shape

(2, 3)

the indexiing operator toys["shape"] returns the correct data, but the attribute-style operation still returns the shape of the dataframe.  You should only use attribute-style accessing in interactive sessions or for read operations. You shouldnt use it for production code or for manipulating data (such as defining new columns)